src/emulator/functionsEmulatorRuntime.ts

```diff
@@ -803,6 +809,56 @@ function rawBodySaver(req: express.Request, res: express.Response, buf: Buffer):
   (req as any).rawBody = buf;
 }
 
+/*
+ Firebase imposes quota limits on response data
+ See https://firebase.google.com/docs/functions/quotas
+*/
+function enforceResponseLimitMiddleware(
+  req: express.Request,
+  res: express.Response,
+  next: express.NextFunction
+): void {
+  let responseSize = 0;
+  const chunks: Buffer[] = [];
+  const originalWrite = res.write.bind(res);
+  const originalEnd = res.end.bind(res);
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  res.write = function (chunk: any) {
+    const buffer = Buffer.from(chunk);
+    responseSize += buffer.length;
+    chunks.push(buffer);
+
+    if (responseSize >= MAX_DATA_LIMIT) {
+      logDebug(`Response size exceeds quota, aborting function execution`);
+      res.emit(
+        "error",
+        new Error(
+          `Response size of ${responseSize} exceeds quota of ${MAX_DATA_LIMIT}, aborting function execution`
+        )
+      );
+      return false;
+    } else {
+      logDebug(`Total response size received: ${responseSize}`);
+      return true;
+    }
+  };
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  res.end = function (chunk: any, encoding?: any, cb?: () => void) {
+    if (chunk) {
```

Comment:
Another thing to consider, apart from the missing unit tests, is that the data limit is not imposed on a direct call to `res.end()` but only on a stream (using `res.write`). This is somewhat deliberate, as I would not expect anyone to directly flush 10MB+ on the response stream, but if the Google Cloud Functions production code does check the single chunk it might be useful to include it here as well.

Reply:
I am not sure about the prod behavior with `res.end()` but this is definitely an improvement either way!